File: Assets/Bundles/Shared/Resources/TS/Entity/Entity.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local GetItemMeta = require("Shared/TS/Item/ItemDefinitions").GetItemMeta
local Network = require("Shared/TS/Network").Network
local NetworkBridge = require("Shared/TS/NetworkBridge").NetworkBridge
local Projectile = require("Shared/TS/Projectile/Projectile").Projectile
local ProjectileUtil = require("Shared/TS/Projectile/ProjectileUtil").ProjectileUtil
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local Signal = require("Shared/TS/Util/Signal").Signal
local TimeUtil = require("Shared/TS/Util/TimeUtil").TimeUtil
local Bin = require("Shared/TS/Util/Bin").Bin
local OnLateUpdate = require("Shared/TS/Util/Timer").OnLateUpdate
local InventoryEntityAnimator = require("Shared/TS/Entity/Animation/InventoryEntityAnimator").InventoryEntityAnimator
local EntitySerializer = require("Shared/TS/Entity/EntitySerializer").EntitySerializer
local WorldAPI = require("Shared/TS/VoxelWorld/WorldAPI").WorldAPI
local EntityReferences
do
	EntityReferences = setmetatable({}, {
		__tostring = function()
			return "EntityReferences"
		end,
	})
	EntityReferences.__index = EntityReferences
	function EntityReferences.new(...)
		local self = setmetatable({}, EntityReferences)
		return self:constructor(...) or self
	end
	function EntityReferences:constructor(ref)
		local boneKey = "Bones"
		local meshKey = "Meshes"
		local colliderKey = "Colliders"
		local vfxKey = "VFX"
		-- Get the meshes
		local meshesCS = ref:GetAllValues(meshKey)
		self.meshes = table.create(meshesCS.Length)
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < meshesCS.Length) then
					break
				end
				self.meshes[i + 1] = meshesCS:GetValue(i)
			end
		end
		self.fpsMesh = ref:GetValue(meshKey, "FirstPerson")
		-- Get the bones
		self.neckBone = ref:GetValue(boneKey, "Neck")
		self.spineBone3 = ref:GetValue(boneKey, "Spine3")
		self.spineBone2 = ref:GetValue(boneKey, "Spine2")
		self.spineBone1 = ref:GetValue(boneKey, "Spine1")
		self.headBone = ref:GetValue(boneKey, "Head")
		self.root = ref:GetValue(boneKey, "Root")
		self.characterCollider = ref:GetValue(colliderKey, "CharacterController")
		self.animationEvents = ref:GetValue(vfxKey, "AnimationEvents")
	end
end
local Entity
do
	Entity = setmetatable({}, {
		__tostring = function()
			return "Entity"
		end,
	})
	Entity.__index = Entity
	function Entity.new(...)
		local self = setmetatable({}, Entity)
		return self:constructor(...) or self
	end
	function Entity:constructor(id, networkObject, clientId)
		self.health = 100
		self.maxHealth = 100
		self.dead = false
		self.destroyed = false
		self.OnHealthChanged = Signal.new()
		self.OnDespawn = Signal.new()
		self.OnPlayerChanged = Signal.new()
		self.id = id
		self.gameObject = networkObject.gameObject
		self.references = EntityReferences.new(self.gameObject:GetComponent("GameObjectReferences"))
		self.model = self.references.root.gameObject
		self.networkObject = networkObject
		self.anim = InventoryEntityAnimator.new(self, self.model:GetComponent("AnimancerComponent"), self.references)
		self.attributes = self.gameObject:GetComponent("EasyAttributes")
		self.accessoryBuilder = self.gameObject:GetComponent("AccessoryBuilder")
		self.dynamicVariables = self.gameObject:GetComponent("DynamicVariables")
		self.ClientId = clientId
		if self.ClientId ~= nil then
			if RunUtil:IsServer() then
				local player = (Flamework.resolveDependency("Bundles/Server/Services/Global/Player/PlayerService@PlayerService")):GetPlayerFromClientId(self.ClientId)
				self:SetPlayer(player)
			else
				local player = (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Player/PlayerController@PlayerController")):GetPlayerFromClientId(self.ClientId)
				if player then
					print("Found Player that controls this entity. (Player=" .. (player.username .. (", Entity=" .. (tostring(self.id) .. ")"))))
				else
					print("Entity is not controlled by any player. (Entity=" .. (tostring(self.id) .. ")"))
				end
				self:SetPlayer(player)
			end
		end
		self.bin = Bin.new()
		self.bin:Connect(OnLateUpdate, function()
			return self:LateUpdate()
		end)
	end
	function Entity:LateUpdate()
		local _result = self.anim
		if _result ~= nil then
			_result:LateUpdate()
		end
	end
	function Entity:SetPlayer(player)
		local oldPlayer = self.player
		self.player = player
		self.OnPlayerChanged:Fire(player, oldPlayer)
	end
	function Entity:GetHealth()
		return self.health
	end
	function Entity:GetMaxHealth()
		return self.maxHealth
	end
	function Entity:GetEntityDriver()
		return self.gameObject:GetComponent("EntityDriver")
	end
	function Entity:SetHealth(health)
		health = math.clamp(health, 0, self.maxHealth)
		if health == self.health then
			return nil
		end
		local oldHealth = self.health
		self.health = health
		self.OnHealthChanged:Fire(health, oldHealth)
		if RunUtil:IsServer() then
			Network.ServerToClient.Entity.SetHealth.Server:FireAllClients(self.id, self.health)
		end
	end
	function Entity:SetMaxHealth(maxHealth)
		self.maxHealth = maxHealth
	end
	function Entity:Destroy()
		self.bin:Clean()
		self.OnDespawn:Fire()
		self.destroyed = true
		self.anim = nil
		if RunUtil:IsServer() then
			Network.ServerToClient.DespawnEntity.Server:FireAllClients(self.id)
			NetworkBridge:Despawn(self.networkObject.gameObject)
		end
	end
	function Entity:IsDestroyed()
		return self.destroyed
	end
	function Entity:Encode()
		return {
			serializer = EntitySerializer.DEFAULT,
			id = self.id,
			clientId = self.ClientId,
			gameObjectId = self.networkObject.ObjectId,
			health = self.health,
			maxHealth = self.maxHealth,
		}
	end
	function Entity:IsPlayerOwned()
		return self.ClientId ~= nil
	end
	function Entity:IsLocalCharacter()
		return RunUtil:IsClient() and self.ClientId == (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Player/PlayerController@PlayerController")).LocalConnection.ClientId
	end
	function Entity:IsAlive()
		return true
	end
	function Entity:FindById(id)
		if RunUtil:IsServer() then
			return (Flamework.resolveDependency("Bundles/Server/Services/Global/Entity/EntityService@EntityService")):GetEntityById(id)
		else
			return (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Entity/EntityController@EntityController")):GetEntityById(id)
		end
	end
	function Entity:FindByClientId(id)
		if RunUtil:IsServer() then
			return (Flamework.resolveDependency("Bundles/Server/Services/Global/Entity/EntityService@EntityService")):GetEntityByClientId(id)
		else
			return (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Entity/EntityController@EntityController")):GetEntityByClientId(id)
		end
	end
	function Entity:FindByCollider(collider)
		local nb = collider.gameObject:GetComponent("NetworkBehaviour")
		if nb ~= nil then
			local split = string.split(nb.name, "_")
			if #split > 0 then
				local entityId = tonumber(split[2])
				if entityId ~= nil then
					return Entity:FindById(entityId)
				end
			end
		end
		return nil
	end
	function Entity:FindByGameObject(gameObject)
		local split = string.split(gameObject.name, "_")
		if #split > 0 then
			local id = tonumber(split[2])
			if id ~= nil then
				return Entity:FindById(id)
			end
		end
		return nil
	end
	function Entity:SendItemAnimationToClients(useIndex, animationMode, exceptClientId)
		if useIndex == nil then
			useIndex = 0
		end
		if animationMode == nil then
			animationMode = 0
		end
		if RunUtil:IsServer() then
			if exceptClientId ~= nil then
				Network.ServerToClient.PlayEntityItemAnimation.Server:FireExcept(exceptClientId, self.id, useIndex, animationMode)
			else
				Network.ServerToClient.PlayEntityItemAnimation.Server:FireAllClients(self.id, useIndex, animationMode)
			end
		else
			error("Trying to send server event (Item Animation) from client")
		end
	end
	function Entity:HasImmunity()
		local immuneUntilTime = self.attributes:GetNumber("immunity")
		if immuneUntilTime ~= nil then
			return TimeUtil:GetServerTime() < immuneUntilTime
		end
		return false
	end
	function Entity:GetLastDamagedTime()
		local _condition = self.attributes:GetNumber("last_damaged")
		if _condition == nil then
			_condition = 0
		end
		return _condition
	end
	function Entity:TimeSinceLastDamaged()
		return Time.time - self:GetLastDamagedTime()
	end
	function Entity:SetLastDamagedTime(time)
		self.attributes:SetAttribute("last_damaged", time)
	end
	function Entity:GrantImmunity(duration)
		local newTime = TimeUtil:GetServerTime() + duration
		local currentTime = self.attributes:GetNumber("immunity")
		if currentTime ~= nil and currentTime > newTime then
			return nil
		end
		self.attributes:SetAttribute("immunity", newTime)
	end
	function Entity:GetState()
		return self:GetEntityDriver():GetState()
	end
	function Entity:GetHeadPosition()
		local state = self:GetState()
		local offset = Vector3.new(0, 1.5, 0)
		if state == 5 then
			offset = Vector3.new(0, 1, 0)
		elseif state == 4 then
			offset = Vector3.new(0, 0.8, 0)
		end
		local _position = self.model.transform.position
		local _offset = offset
		return _position + _offset
	end
	function Entity:GetMiddlePosition()
		local _position = self.model.transform.position
		local _vector3 = Vector3.new(0, 1.5, 0)
		return _position + _vector3
	end
	function Entity:LocalOffsetToWorldPoint(localOffset)
		local worldDir = self.model.transform:TransformDirection(localOffset)
		local worldPoint = self:GetMiddlePosition() + worldDir
		return worldPoint
	end
	function Entity:GetDisplayName()
		if self.player then
			return self.player.username
		end
		return "entity_" .. tostring(self.id)
	end
	function Entity:Kill()
		if self.dead then
			return nil
		end
		self.dead = true
	end
	function Entity:IsDead()
		return self.dead
	end
	function Entity:GetBlockBelowMeta()
		return WorldAPI:GetMainWorld():GetBlockBelowMeta(self.model.transform.position)
	end
	function Entity:GetAccessoryGameObjects(slot)
		return self:PushToArray(self.accessoryBuilder:GetAccessories(slot))
	end
	function Entity:GetAccessoryMeshes(slot)
		return self:PushToArray(self.accessoryBuilder:GetAccessoryMeshes(slot))
	end
	function Entity:PushToArray(array)
		local results = {}
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < array.Length) then
					break
				end
				local _results = results
				local _arg0 = array:GetValue(i)
				table.insert(_results, _arg0)
			end
		end
		return results
	end
	function Entity:LaunchProjectile(itemType, velocity)
		local itemMeta = GetItemMeta(itemType)
		if not itemMeta.Ammo then
			return error("Tried to launch item that wasn't a projectile: " .. itemType)
		end
		local firstPerson = false
		if self:IsLocalCharacter() then
			firstPerson = (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController")):IsFirstPerson()
		end
		local launchPos = ProjectileUtil:GetLaunchPosition(self, firstPerson)
		local projectilePath = "Shared/Resources/Prefabs/Projectiles/Ammo/" .. (itemType .. ".prefab")
		local projectileLauncher = self.gameObject:GetComponent("ProjectileLauncher")
		local easyProjectile = projectileLauncher:ClientFire(projectilePath, itemMeta.ID, launchPos, velocity, itemMeta.Ammo.gravity, 0)
		local projectile = Projectile.new(easyProjectile, itemType, self)
		if RunUtil:IsClient() then
			local clientSignals = TS.Promise.new(function(resolve)
				resolve(require("Client/TS/ClientSignals"))
			end):expect().ClientSignals
			local ProjectileLaunchedClientSignal = TS.Promise.new(function(resolve)
				resolve(require("Client/TS/Controllers/Global/Damage/Projectile/ProjectileLaunchedClientSignal"))
			end):expect().ProjectileLaunchedClientSignal
			clientSignals.ProjectileLaunched:Fire(ProjectileLaunchedClientSignal.new(projectile))
		end
	end
end
return {
	EntityReferences = EntityReferences,
	Entity = Entity,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Entity/Entity.ts
import { Dependency } from "@easy-games/flamework-core";
import { LocalEntityController } from "Client/Controllers/Global/Character/LocalEntityController";
import { EntityController } from "Client/Controllers/Global/Entity/EntityController";
import { PlayerController } from "Client/Controllers/Global/Player/PlayerController";
import { EntityService } from "Server/Services/Global/Entity/EntityService";
import { PlayerService } from "Server/Services/Global/Player/PlayerService";
import { GameObjectUtil } from "Shared/GameObjectBridge";
import { ItemType } from "Shared/Item/ItemType";
import { Network } from "Shared/Network";
import { Player } from "Shared/Player/Player";
import { Projectile } from "Shared/Projectile/Projectile";
import { ProgressBarGraphics } from "Shared/UI/ProgressBarGraphics";
import { NetworkUtil } from "Shared/Util/NetworkUtil";
import { RunUtil } from "Shared/Util/RunUtil";
import { Signal } from "Shared/Util/Signal";
import { TimeUtil } from "Shared/Util/TimeUtil";
import { AudioManager } from "../Audio/AudioManager";
import { BlockMeta } from "../Item/ItemMeta";
import { ItemUtil } from "../Item/ItemUtil";
import { BundleReferenceManager } from "../Util/BundleReferenceManager";
import { BundleGroupNames, Bundle_Entity, Bundle_Entity_Movement } from "../Util/ReferenceManagerResources";
import { WorldAPI } from "../VoxelWorld/WorldAPI";
import { InventoryEntityAnimator, ItemPlayMode } from "./Animation/InventoryEntityAnimator";
import { EntitySerializer } from "./EntitySerializer";

export interface EntityDto {
	serializer: EntitySerializer;
	id: number;
	/** Fish-net ObjectId */
	nobId: number;
	clientId?: number;
	health: number;
	maxHealth: number;
	displayName: string;
	healthbar?: boolean;
}

export class EntityReferences {
	meshes: Array<Renderer>;
	fpsMesh: Renderer;
	neckBone: Transform;
	headBone: Transform;
	spineBone1: Transform;
	spineBone2: Transform;
	spineBone3: Transform;
	root: Transform;
	characterCollider: Collider;
	animationEvents: EntityAnimationEvents;
	jumpSound: AudioClip | undefined;
	slideSound: AudioClip | undefined;
	landSound: AudioClip | undefined;

	constructor(ref: GameObjectReferences) {
		let boneKey = "Bones";
		let meshKey = "Meshes";
		let colliderKey = "Colliders";
		let vfxKey = "VFX";

		//Get the meshes
		let meshesCS: CSArray<Renderer> = ref.GetAllValues<Renderer>(meshKey);
		this.meshes = table.create(meshesCS.Length);
		for (let i = 0; i < meshesCS.Length; i++) {
			this.meshes[i] = meshesCS.GetValue(i);
		}
		this.fpsMesh = ref.GetValue<Renderer>(meshKey, "FirstPerson");
		//Get the bones
		this.neckBone = ref.GetValue<Transform>(boneKey, "Neck");
		this.spineBone3 = ref.GetValue<Transform>(boneKey, "Spine3");
		this.spineBone2 = ref.GetValue<Transform>(boneKey, "Spine2");
		this.spineBone1 = ref.GetValue<Transform>(boneKey, "Spine1");
		this.headBone = ref.GetValue<Transform>(boneKey, "Head");
		this.root = ref.GetValue<Transform>(boneKey, "Root");

		this.characterCollider = ref.GetValue<Collider>(colliderKey, "CharacterController");

		this.animationEvents = ref.GetValue<EntityAnimationEvents>(vfxKey, "AnimationEvents");

		/*this.jumpSound = AudioManager.LoadFullPathAudioClip(BundleReferenceManager.GetPathForResource(
			BundleGroupNames.Entity,
			Bundle_Entity.Movement,
			Bundle_Entity_Movement.JumpSFX,
		));*/

		this.slideSound = AudioManager.LoadFullPathAudioClip(
			BundleReferenceManager.GetPathForResource(
				BundleGroupNames.Entity,
				Bundle_Entity.Movement,
				Bundle_Entity_Movement.SlideSFX,
			),
		);

		/*this.landSound = AudioManager.LoadFullPathAudioClip(
			BundleReferenceManager.GetPathForResource(
				BundleGroupNames.Entity,
				Bundle_Entity.Movement,
				Bundle_Entity_Movement.LandSFX,
			),
		);*/
	}
}

export class Entity {
	/** Entity's unique id. */
	public readonly id: number;
	public readonly gameObject: GameObject;
	public readonly networkObject: NetworkObject;
	public readonly entityDriver: EntityDriver;
	public readonly model: GameObject;
	public readonly attributes: EasyAttributes;
	public anim?: InventoryEntityAnimator;
	public readonly references: EntityReferences;
	public readonly accessoryBuilder: AccessoryBuilder;

	public player: Player | undefined;

	/**
	 * The connection ID of whoever is controlling this entity.
	 * Only exists if this entity is attached to a player.
	 *
	 * **This should NOT be used to uniquely identify an entity.**
	 */
	public readonly ClientId?: number;
	private health = 100;
	private maxHealth = 100;
	private dead = false;
	private destroyed = false;
	private displayName: string;
	private healthbarEnabled = false;
	private healthbar?: ProgressBarGraphics;

	public readonly OnHealthChanged = new Signal<[newHealth: number, oldHealth: number]>();
	public readonly OnDespawn = new Signal<void>();
	public readonly OnPlayerChanged = new Signal<[newPlayer: Player | undefined, oldPlayer: Player | undefined]>();
	public readonly OnAdjustMove = new Signal<[moveModifier: MoveModifier]>();
	public readonly OnDisplayNameChanged = new Signal<[displayName: string]>();

	constructor(id: number, networkObject: NetworkObject, clientId: number | undefined) {
		this.id = id;
		this.gameObject = networkObject.gameObject;
		this.references = new EntityReferences(this.gameObject.GetComponent<GameObjectReferences>());
		this.model = this.references.root.gameObject;
		this.entityDriver = this.gameObject.GetComponent<EntityDriver>();
		this.networkObject = networkObject;
		this.anim = new InventoryEntityAnimator(this, this.model.GetComponent<AnimancerComponent>(), this.references);
		this.attributes = this.gameObject.GetComponent<EasyAttributes>();
		this.accessoryBuilder = this.gameObject.GetComponent<AccessoryBuilder>();
		this.ClientId = clientId;
		if (this.ClientId !== undefined) {
			if (RunUtil.IsServer()) {
				const player = Dependency<PlayerService>().GetPlayerFromClientId(this.ClientId);
				this.SetPlayer(player);
			} else {
				const player = Dependency<PlayerController>().GetPlayerFromClientId(this.ClientId);
				if (player) {
					print(`Found Player that controls this entity. (Player=${player.username}, Entity=${this.id})`);
				} else {
					print(`Entity is not controlled by any player. (Entity=${this.id})`);
				}
				this.SetPlayer(player);
			}
		}
		if (this.player) {
			this.displayName = this.player.username;
		} else {
			this.displayName = `entity_${this.id}`;
		}

		this.entityDriver.OnImpactWithGround((velocity) => {
			this.anim?.PlayFootstepSound();
		});

		this.entityDriver.OnAdjustMove((moveModifier) => {
			this.OnAdjustMove.Fire(moveModifier);
		});
	}

	public AddHealthbar(): void {
		if (RunUtil.IsServer()) {
			this.healthbarEnabled = true;
			Network.ServerToClient.Entity.AddHealthbar.Server.FireAllClients(this.id);
			return;
		}
		if (this.IsLocalCharacter()) return;

		const prefab = AssetBridge.LoadAsset<Object>("Client/Resources/Prefabs/EntityHealthbar.prefab");
		const healthbarGO = GameObjectUtil.InstantiateIn(prefab, this.model.transform);
		const transform = healthbarGO.transform;
		transform.localPosition = new Vector3(0, 2.2, 0);
		this.healthbar = new ProgressBarGraphics(transform.GetChild(0));

		this.healthbar.SetValue(this.health / this.maxHealth);
		this.healthbarEnabled = true;
	}

	public SetPlayer(player: Player | undefined): void {
		const oldPlayer = this.player;
		this.player = player;
		this.OnPlayerChanged.Fire(player, oldPlayer);
	}

	public SetDisplayName(displayName: string) {
		this.displayName = displayName;
		this.OnDisplayNameChanged.Fire(displayName);
		if (RunUtil.IsServer()) {
			Network.ServerToClient.Entity.SetDisplayName.Server.FireAllClients(this.id, displayName);
		}
	}

	public GetHealth(): number {
		return this.health;
	}

	public GetMaxHealth(): number {
		return this.maxHealth;
	}

	public GetEntityDriver(): EntityDriver {
		return this.entityDriver;
	}

	public SetHealth(health: number): void {
		health = math.clamp(health, 0, this.maxHealth);
		if (health === this.health) return;
		const oldHealth = this.health;
		this.health = health;
		this.OnHealthChanged.Fire(health, oldHealth);
		this.healthbar?.SetValue(this.health / this.maxHealth);

		if (RunUtil.IsServer()) {
			Network.ServerToClient.Entity.SetHealth.Server.FireAllClients(this.id, this.health);
		}
	}

	public SetMaxHealth(maxHealth: number): void {
		this.maxHealth = maxHealth;
	}

	/**
	 * It is recommended to use EntityService.DespawnEntity() instead of this.
	 */
	public Destroy(): void {
		this.OnDespawn.Fire();
		this.destroyed = true;

		delete this.anim;

		if (this.player) {
			this.player.SetCharacter(undefined);
		}

		if (RunUtil.IsServer()) {
			Network.ServerToClient.DespawnEntity.Server.FireAllClients(this.id);
			NetworkUtil.Despawn(this.networkObject.gameObject);
		}
	}

	public IsDestroyed(): boolean {
		return this.destroyed;
	}

	public Encode(): EntityDto {
		return {
			serializer: EntitySerializer.DEFAULT,
			id: this.id,
			clientId: this.ClientId,
			nobId: this.networkObject.ObjectId,
			health: this.health,
			maxHealth: this.maxHealth,
			displayName: this.displayName,
			healthbar: this.healthbarEnabled || undefined,
		};
	}

	public IsPlayerOwned(): boolean {
		return this.ClientId !== undefined;
	}

	public IsLocalCharacter(): boolean {
		if (!RunUtil.IsClient()) {
			return false;
		} else {
			return this.ClientId === Dependency<PlayerController>().LocalConnection.ClientId;
		}
	}

	public IsAlive(): boolean {
		return !this.IsDead();
	}

	public static FindById(id: number): Entity | undefined {
		if (RunUtil.IsServer()) {
			return Dependency<EntityService>().GetEntityById(id);
		} else {
			return Dependency<EntityController>().GetEntityById(id);
		}
	}

	public static async WaitForId(id: number): Promise<Entity | undefined> {
		if (RunUtil.IsServer()) {
			return this.FindById(id);
		} else {
			return await Dependency<EntityController>().WaitForId(id);
		}
	}

	public static FindByClientId(id: number): Entity | undefined {
		if (RunUtil.IsServer()) {
			return Dependency<EntityService>().GetEntityByClientId(id);
		} else {
			return Dependency<EntityController>().GetEntityByClientId(id);
		}
	}

	public static FindByCollider(collider: Collider): Entity | undefined {
		let nb = collider.gameObject.GetComponent<NetworkBehaviour>();
		if (nb === undefined) {
			nb = collider.transform.parent.gameObject.GetComponent<NetworkBehaviour>();
		}

		if (nb !== undefined) {
			const split = nb.name.split("_");
			if (split.size() > 0) {
				const entityId = tonumber(split[1])!;
				if (entityId !== undefined) {
					return Entity.FindById(entityId);
				}
			}
		}

		return undefined;
	}

	public static FindByGameObject(gameObject: GameObject): Entity | undefined {
		const split = gameObject.name.split("_");
		if (split.size() > 0) {
			const id = tonumber(split[1]);
			if (id !== undefined) {
				return Entity.FindById(id);
			}
		}
		return undefined;
	}

	/*public PlayAnimation(
		animationId: EntityAnimationId,
		config?: {
			layer?: number;
			exceptClientId?: number;
		},
	) {
		if (RunUtil.IsServer()) {
			if (config?.exceptClientId !== undefined) {
				Network.ServerToClient.PlayEntityAnimation.Server.FireExcept(
					config.exceptClientId,
					this.Id,
					animationId,
					config?.layer,
				);
			} else {
				Network.ServerToClient.PlayEntityAnimation.Server.FireAllClients(this.Id, animationId, config?.layer);
			}
		}
		this.anim.PlayAnimation(animationId, config?.layer ?? 10);
	}*/

	public SendItemAnimationToClients(useIndex = 0, animationMode: ItemPlayMode = 0, exceptClientId?: number) {
		if (RunUtil.IsServer()) {
			if (exceptClientId !== undefined) {
				Network.ServerToClient.PlayEntityItemAnimation.Server.FireExcept(
					exceptClientId,
					this.id,
					useIndex,
					animationMode,
				);
			} else {
				Network.ServerToClient.PlayEntityItemAnimation.Server.FireAllClients(this.id, useIndex, animationMode);
			}
		} else {
			error("Trying to send server event (Item Animation) from client");
		}
	}

	public HasImmunity(): boolean {
		let immuneUntilTime = this.attributes.GetNumber("immunity");
		if (immuneUntilTime !== undefined) {
			return TimeUtil.GetServerTime() < immuneUntilTime;
		}
		return false;
	}

	public GetImmuneUntilTime(): number {
		return this.attributes.GetNumber("immunity") ?? 0;
	}

	public GetLastDamagedTime(): number {
		return this.attributes.GetNumber("last_damaged") ?? 0;
	}

	public TimeSinceLastDamaged(): number {
		return Time.time - this.GetLastDamagedTime();
	}

	public SetLastDamagedTime(time: number): void {
		this.attributes.SetAttribute("last_damaged", time);
	}

	public GrantImmunity(duration: number): void {
		let newTime = TimeUtil.GetServerTime() + duration;

		let currentTime = this.attributes.GetNumber("immunity");
		if (currentTime !== undefined && currentTime > newTime) {
			return;
		}

		this.attributes.SetAttribute("immunity", newTime);
	}

	public GetState(): EntityState {
		return this.GetEntityDriver().GetState();
	}

	public GetHeadPosition(): Vector3 {
		const offset = this.GetHeadOffset();
		return this.model.transform.position.add(offset);
	}

	public GetHeadOffset(): Vector3 {
		const state = this.GetState();
		let offset = new Vector3(0, 1.7, 0);
		if (state === EntityState.Crouching) {
			offset = new Vector3(0, 1, 0);
		} else if (state === EntityState.Sliding) {
			offset = new Vector3(0, 0.8, 0);
		}
		return offset;
	}

	public GetMiddlePosition(): Vector3 {
		return this.model.transform.position.add(new Vector3(0, 0.9, 0));
	}

	public LocalOffsetToWorldPoint(localOffset: Vector3) {
		const worldDir = this.model.transform.TransformDirection(localOffset);
		const worldPoint = this.GetMiddlePosition().add(worldDir);
		return worldPoint;
	}

	public GetDisplayName(): string {
		return this.displayName;
	}

	public Kill(): void {
		if (this.dead) return;
		this.dead = true;
	}

	public IsDead(): boolean {
		return this.dead;
	}

	public GetBlockBelowMeta(): BlockMeta | undefined {
		return WorldAPI.GetMainWorld().GetBlockBelowMeta(this.model.transform.position);
	}

	public GetAccessoryMeshes(slot: AccessorySlot): Renderer[] {
		return this.PushToArray(this.accessoryBuilder.GetAccessoryMeshes(slot));
	}

	private PushToArray<T>(array: CSArray<T>): T[] {
		let results = new Array<any>();
		for (let i = 0; i < array.Length; i++) {
			results.push(array.GetValue(i));
		}
		return results;
	}

	public LaunchProjectile(itemType: ItemType, launchPos: Vector3, velocity: Vector3): EasyProjectile | undefined {
		const itemMeta = ItemUtil.GetItemMeta(itemType);
		if (!itemMeta.Ammo) {
			return error("Tried to launch item that wasn't a projectile: " + itemType);
		}
		let firstPerson = false;
		if (this.IsLocalCharacter()) {
			firstPerson = Dependency<LocalEntityController>().IsFirstPerson();
		}
		const projectilePath = `Shared/Resources/Prefabs/Projectiles/Ammo/${itemType}.prefab`;
		const projectileLauncher = this.gameObject.GetComponent<ProjectileLauncher>();

		const easyProjectile = projectileLauncher.ClientFire(
			projectilePath,
			itemMeta.ID,
			launchPos,
			velocity,
			itemMeta.Ammo.gravity,
			0,
		);
		const projectile = new Projectile(easyProjectile, itemType, this);

		if (RunUtil.IsClient()) {
			const clientSignals = import("Client/ClientSignals").expect().ClientSignals;
			const ProjectileLaunchedClientSignal = import(
				"Client/Controllers/Global/Damage/Projectile/ProjectileLaunchedClientSignal"
			).expect().ProjectileLaunchedClientSignal;

			clientSignals.ProjectileLaunched.Fire(new ProjectileLaunchedClientSignal(projectile));
		}
	}

	public HasHealthbar(): boolean {
		return this.healthbarEnabled;
	}
}
